### ISSUE TITLE
regression: assert_eq!(old_config, Some(new_config)) in src/compute/src/server.rs:439

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#13914]
### What version of Materialize are you using?

environmentd v0.26.1-dev (550670b32)

### How did you install Materialize?

Docker image

### What is the issue?

Asserts as shown below with this test that restarts environmentd:

```
cd test/platform-checks
./mzcompose run default --scenario=RestartEnvironmentdStoraged
```

The assert was introduced here:

```
commit 33143731de58e6a55d07466b79a0de9eeef70608
Author: Moritz Hoffmann <mh@materialize.com>
Date:   Tue Jul 26 14:56:19 2022 +0200

    Revert "Permit hydration of an uninitialized cluster"
    
    This reverts commit 625c4888772d71a10c14b6b1f4e90eb24203a028.
```

### Relevant log output

```source
thread 'timely:work-0' panicked at 'assertion failed: `(left == right)`
  left: `Some(InstanceConfig { replica_id: 2, logging: Some(LoggingConfig { granularity_ns: 1000000000, active_logs: {Timely(Operates): System(329), Timely(Channels): System(328), Timely(Elapsed): System(338), Timely(Histogram): System(339), Timely(Addresses): System(330), Timely(Parks): System(340), Timely(MessagesSent): System(335), Timely(MessagesReceived): System(334), Timely(Reachability): System(331), Differential(ArrangementBatches): System(326), Differential(ArrangementRecords): System(327), Differential(Sharing): System(325), Compute(DataflowCurrent): System(332), Compute(DataflowDependency): System(333), Compute(FrontierCurrent): System(341), Compute(PeekCurrent): System(336), Compute(PeekDuration): System(337)}, log_logging: false, sink_logs: {Timely(Operates): (System(359), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(58b79618-f4a6-4746-b406-02cd7326d4ac), data_shard: ShardId(dd772da7-5037-420a-9c56-2567e8c05c77), status_shard: ShardId(22b0c801-890c-4441-90d5-4ee2b28121a3) }), Timely(Channels): (System(360), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(b06dad91-2b87-4216-a302-f3563b356e08), data_shard: ShardId(de0c87d4-a227-4972-8ddc-72be2db8722b), status_shard: ShardId(60604fc2-1592-48a5-9f7b-d6a748805318) }), Timely(Elapsed): (System(361), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(0f143424-3afc-401a-8271-3f507705ae6a), data_shard: ShardId(55b6054f-9457-492a-bc70-4a088f266bf6), status_shard: ShardId(aa392abb-bf54-42bb-8485-b9e989c1c323) }), Timely(Histogram): (System(362), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(971265ea-9ce4-4efb-95e7-ec8ea178b6c8), data_shard: ShardId(e91c28e1-4890-465d-9e8f-a2144233a160), status_shard: ShardId(aa60709f-b135-4260-9571-bdc937a88554) }), Timely(Addresses): (System(363), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(0f8762a4-0c8c-4a8b-88e2-896963de51ea), data_shard: ShardId(df151e70-5a6a-41d2-8bd7-4e491569df58), status_shard: ShardId(27303f11-0066-47c7-ba32-975c1301fd01) }), Timely(Parks): (System(364), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(017f52d9-f4ba-40bb-adde-fe05c5fab5ce), data_shard: ShardId(8395ead1-75fc-412f-9c32-f34c8f76c556), status_shard: ShardId(de295ffe-41c0-4cdb-98f1-79b4cee9f92e) }), Timely(MessagesSent): (System(365), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(c3834ba8-6316-453f-ba4f-387d46c15ebb), data_shard: ShardId(df7cdb9e-5033-4a8f-8a8c-9d4ffbd3806c), status_shard: ShardId(959d3528-d510-47e1-b637-32c38c4ac87c) }), Timely(MessagesReceived): (System(366), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(5170cc0b-a71d-4584-9ea1-617b1cc390dc), data_shard: ShardId(d8ef8b49-c6cc-4219-b6b7-b217d97dda41), status_shard: ShardId(37a3c88c-d781-4a60-86a4-f462b7907745) }), Timely(Reachability): (System(367), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(b5cc76a1-7a3e-4809-bfcb-a0ac6e9a12af), data_shard: ShardId(5459ec9b-d7e1-4d79-8738-0d92d30f4766), status_shard: ShardId(14365064-09c7-47b9-bb0c-95597a7f6710) }), Differential(ArrangementBatches): (System(368), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(f31e5e7d-3e0e-41eb-9d3e-9f9d2de9ab5b), data_shard: ShardId(475ca9e6-122d-4260-9cc2-e0175c1c1401), status_shard: ShardId(ad850336-afda-4c73-9045-33bc53ef1dc0) }), Differential(ArrangementRecords): (System(369), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(ec71732f-bf74-49a1-b4f8-75b23f035a68), data_shard: ShardId(a7de70f4-ae8c-4411-a956-40aa942c51f3), status_shard: ShardId(f4288ee9-a17d-43d1-b456-c43f2a4c65c2) }), Differential(Sharing): (System(370), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(efe2c3fe-88ed-4685-a0ea-be0c7ba4fb4d), data_shard: ShardId(22b890a2-c890-443f-888f-419519c913f1), status_shard: ShardId(5a54ef42-80fe-458c-a68c-23bb989b6d28) }), Compute(DataflowCurrent): (System(371), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(622651c7-ed42-4023-8743-49a7c7bf9a38), data_shard: ShardId(78f278bf-b169-4b46-b95c-6d3a43daaaef), status_shard: ShardId(190aae74-6f56-4c15-95c4-7a06cccb0f36) }), Compute(DataflowDependency): (System(372), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(6eb10800-15a8-413b-a00e-ad5f394e74ea), data_shard: ShardId(a3d96f84-0f7e-491d-980b-3a818d14af6d), status_shard: ShardId(6da7018d-bdce-4134-a5bb-d164f7f9a687) }), Compute(FrontierCurrent): (System(373), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(36b66827-a2c5-4c8c-87a4-217933b90db6), data_shard: ShardId(17280c7c-79a5-4408-a12d-9f427d615ba8), status_shard: ShardId(97d49523-c21a-4cad-83d8-624419872c0e) }), Compute(PeekCurrent): (System(374), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(d6dc9300-3b61-43fa-8c93-d7f53aa9e281), data_shard: ShardId(fcbe47f0-2255-4430-8746-042960b447f0), status_shard: ShardId(cbc8d717-8a37-4177-957d-622497db87ce) }), Compute(PeekDuration): (System(375), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(a625e010-2fda-43a9-b74c-90d886a30d4b), data_shard: ShardId(80993774-177f-477c-9b11-9cd33fd9a141), status_shard: ShardId(a2dda8e4-33df-4eb8-b7bf-a39be18b5b1c) })} }) })`,
 right: `Some(InstanceConfig { replica_id: 2, logging: Some(LoggingConfig { granularity_ns: 1000000000, active_logs: {Timely(Operates): System(329), Timely(Channels): System(328), Timely(Elapsed): System(338), Timely(Histogram): System(339), Timely(Addresses): System(330), Timely(Parks): System(340), Timely(MessagesSent): System(335), Timely(MessagesReceived): System(334), Timely(Reachability): System(331), Differential(ArrangementBatches): System(326), Differential(ArrangementRecords): System(327), Differential(Sharing): System(325), Compute(DataflowCurrent): System(332), Compute(DataflowDependency): System(333), Compute(FrontierCurrent): System(341), Compute(PeekCurrent): System(336), Compute(PeekDuration): System(337)}, log_logging: false, sink_logs: {Timely(Operates): (System(359), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(58b79618-f4a6-4746-b406-02cd7326d4ac), data_shard: ShardId(dd772da7-5037-420a-9c56-2567e8c05c77), status_shard: ShardId(14178365-f956-4372-aa8a-5c378da661aa) }), Timely(Channels): (System(360), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(b06dad91-2b87-4216-a302-f3563b356e08), data_shard: ShardId(de0c87d4-a227-4972-8ddc-72be2db8722b), status_shard: ShardId(4ecc8e5b-c359-426d-9913-b975e24c977c) }), Timely(Elapsed): (System(361), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(0f143424-3afc-401a-8271-3f507705ae6a), data_shard: ShardId(55b6054f-9457-492a-bc70-4a088f266bf6), status_shard: ShardId(f15fdb14-7bd2-472e-a136-2a4bca7e6bad) }), Timely(Histogram): (System(362), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(971265ea-9ce4-4efb-95e7-ec8ea178b6c8), data_shard: ShardId(e91c28e1-4890-465d-9e8f-a2144233a160), status_shard: ShardId(56a3561e-f03b-4d0c-8d7b-80e7a4d37e9d) }), Timely(Addresses): (System(363), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(0f8762a4-0c8c-4a8b-88e2-896963de51ea), data_shard: ShardId(df151e70-5a6a-41d2-8bd7-4e491569df58), status_shard: ShardId(e1dbdfa0-b296-41e7-a1de-de529f344ecf) }), Timely(Parks): (System(364), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(017f52d9-f4ba-40bb-adde-fe05c5fab5ce), data_shard: ShardId(8395ead1-75fc-412f-9c32-f34c8f76c556), status_shard: ShardId(96be56d6-0c0b-487f-87e5-da1a7484646a) }), Timely(MessagesSent): (System(365), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(c3834ba8-6316-453f-ba4f-387d46c15ebb), data_shard: ShardId(df7cdb9e-5033-4a8f-8a8c-9d4ffbd3806c), status_shard: ShardId(e0aa75d6-05ab-440f-b5d1-f4cb855d8ef0) }), Timely(MessagesReceived): (System(366), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(5170cc0b-a71d-4584-9ea1-617b1cc390dc), data_shard: ShardId(d8ef8b49-c6cc-4219-b6b7-b217d97dda41), status_shard: ShardId(514ad36b-c9bf-425e-a7aa-f2fd145a1c7f) }), Timely(Reachability): (System(367), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(b5cc76a1-7a3e-4809-bfcb-a0ac6e9a12af), data_shard: ShardId(5459ec9b-d7e1-4d79-8738-0d92d30f4766), status_shard: ShardId(d9adeca2-0256-4eb8-8085-2199c66f0fca) }), Differential(ArrangementBatches): (System(368), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(f31e5e7d-3e0e-41eb-9d3e-9f9d2de9ab5b), data_shard: ShardId(475ca9e6-122d-4260-9cc2-e0175c1c1401), status_shard: ShardId(5393a4ef-97e4-433b-9b6a-fe55668a893e) }), Differential(ArrangementRecords): (System(369), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(ec71732f-bf74-49a1-b4f8-75b23f035a68), data_shard: ShardId(a7de70f4-ae8c-4411-a956-40aa942c51f3), status_shard: ShardId(995a1c37-2d25-4a95-a1f2-b56ac04b727f) }), Differential(Sharing): (System(370), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(efe2c3fe-88ed-4685-a0ea-be0c7ba4fb4d), data_shard: ShardId(22b890a2-c890-443f-888f-419519c913f1), status_shard: ShardId(8521ffd9-ce97-4994-9e4c-34804a654470) }), Compute(DataflowCurrent): (System(371), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(622651c7-ed42-4023-8743-49a7c7bf9a38), data_shard: ShardId(78f278bf-b169-4b46-b95c-6d3a43daaaef), status_shard: ShardId(3b40e26d-4653-4fed-bc5e-f6da76e6e549) }), Compute(DataflowDependency): (System(372), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(6eb10800-15a8-413b-a00e-ad5f394e74ea), data_shard: ShardId(a3d96f84-0f7e-491d-980b-3a818d14af6d), status_shard: ShardId(4fdd3105-50c9-444c-b677-f0ad6894a605) }), Compute(FrontierCurrent): (System(373), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(36b66827-a2c5-4c8c-87a4-217933b90db6), data_shard: ShardId(17280c7c-79a5-4408-a12d-9f427d615ba8), status_shard: ShardId(9265b609-18fd-42e4-8ed3-5333a9325c26) }), Compute(PeekCurrent): (System(374), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(d6dc9300-3b61-43fa-8c93-d7f53aa9e281), data_shard: ShardId(fcbe47f0-2255-4430-8746-042960b447f0), status_shard: ShardId(7ee42f90-9e87-4948-a58d-de075cb4974f) }), Compute(PeekDuration): (System(375), CollectionMetadata { persist_location: PersistLocation { blob_uri: "file:///mzdata/persist/blob", consensus_uri: "postgresql://postgres:postgres@postgres-stash:5432?options=--search_path=consensus" }, remap_shard: ShardId(a625e010-2fda-43a9-b74c-90d886a30d4b), data_shard: ShardId(80993774-177f-477c-9b11-9cd33fd9a141), status_shard: ShardId(3d83abb8-6e1c-48c8-a991-52b21924a63f) })} }) })`', src/compute/src/server.rs:439:25
stack backtrace:
   0: rust_begin_unwind
             at ./rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at ./rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:142:14
   2: core::panicking::assert_failed_inner
   3: core::panicking::assert_failed
             at ./rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/core/src/panicking.rs:181:5
   4: mz_compute::server::Worker<A>::reconcile
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0fba999a134df34b0-1/materialize/tests/src/compute/src/server.rs:439:25
   5: mz_compute::server::Worker<A>::run_client
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0fba999a134df34b0-1/materialize/tests/src/compute/src/server.rs:193:25
   6: mz_compute::server::Worker<A>::run
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0fba999a134df34b0-1/materialize/tests/src/compute/src/server.rs:185:33
   7: mz_compute::server::serve::{{closure}}
             at ./var/lib/buildkite-agent/builds/buildkite-builders-d43b1b5-i-0fba999a134df34b0-1/materialize/tests/src/compute/src/server.rs:113:13
   8: timely::execute::execute_from::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/ee65cdc/timely/src/execute.rs:388:22
   9: timely_communication::initialize::initialize_from::{{closure}}
             at ./cargo/git/checkouts/timely-dataflow-4c0cc365061cd263/ee65cdc/communication/src/initialize.rs:316:33
```

**Comment by antiguru** [comment]
This seems like an actual bug:
* old: `status_shard: ShardId(a2dda8e4-33df-4eb8-b7bf-a39be18b5b1c)`
* new: `status_shard: ShardId(3d83abb8-6e1c-48c8-a991-52b21924a63f)`
This means the configuration for the instance changed and the only thing it can do is crash.

**Comment by frankmcsherry** [comment]
I think @andrioni and @benesch know what `status_shard` is and can perhaps opine on whether it should change when a replica is reconnected to!

**Comment by andrioni** [comment]
`status_shard` used to be persisted, but this was changed in #13806: after this, it can indeed vary for objects that are not sources (the adapter guarantees that sources have a stable assignment, but everything else gets a random `ShardId` every `create_collections` call), but it's also not used in those cases. I guess the best fix here would probably to turn it into an `Option` and just not fill it at all for non-sources. I can tackle it later today or tomorrow.

**Comment by philip-stoev** [comment]
Bisection failed unfortunately, here are a few failing commits:

git bisect bad 54ba3052f58ed083f046358976570890f3271c3b
git bisect bad 9af5d3d4452d88ec71ae325ea929666372d17861
git bisect bad 56404052d5fbf57372c12db9ce4da883c0ca3061
git bisect bad bc3ea9f929a7df907057f3879b31e85890bb0967

**Comment by benesch** [comment]
> I guess the best fix here would probably to turn it into an Option and just not fill it at all for non-sources. I can tackle it later today or tomorrow.

Oops, yup, that sounds right.

**Comment by uce** [comment]
Fixed in #13917.
